roachtest: tpcc/multiregion/survive=zone/chaos=true failed

roachtest.tpcc/multiregion/survive=zone/chaos=true [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4090651&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4090651&tab=artifacts#/tpcc/multiregion/survive=zone/chaos=true) on master @ [fdd672fed84323279f7070127f36cdbdc5c26b03](https://github.com/cockroachdb/cockroach/commits/fdd672fed84323279f7070127f36cdbdc5c26b03):


```
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkUptime
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:57
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).listen.func1
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:256
		  |   | runtime.goexit
		  |   | 	/usr/local/go/src/runtime/asm_amd64.s:1581
		  | Wraps: (4) expected <=200 errors, found from 1093754.000000, to 1094224.000000
		  | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
		Wraps: (72) secondary error attachment
		  | error at from 2022-01-11T15:57:43Z, to 2022-01-11T16:02:23Z on metric workload_tpcc_newOrder_error_total{instance="35.185.40.99:2112"}: expected <=200 errors, found from 323.000000, to 845.000000
		  | (1) attached stack trace
		  |   -- stack trace:
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:208
		  |   | [...repeated from below...]
		  | Wraps: (2) error at from 2022-01-11T15:57:43Z, to 2022-01-11T16:02:23Z on metric workload_tpcc_newOrder_error_total{instance="35.185.40.99:2112"}
		  | Wraps: (3) attached stack trace
		  |   -- stack trace:
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkUptime.func2
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:79
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:207
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkUptime
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:57
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).listen.func1
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:283
		  |   | runtime.goexit
		  |   | 	/usr/local/go/src/runtime/asm_amd64.s:1581
		  | Wraps: (4) expected <=200 errors, found from 323.000000, to 845.000000
		  | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
		Wraps: (73) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:208
		  | [...repeated from below...]
		Wraps: (74) error at from 2022-01-11T15:57:43Z, to 2022-01-11T16:02:23Z on metric workload_tpcc_newOrder_error_total{instance="35.185.40.99:2111"}
		Wraps: (75) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkUptime.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:79
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:207
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkUptime
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:57
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).listen.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:283
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1581
		Wraps: (76) expected <=200 errors, found from 340.000000, to 851.000000
		Error types: (1) *secondary.withSecondaryError (2) *secondary.withSecondaryError (3) *secondary.withSecondaryError (4) *secondary.withSecondaryError (5) *secondary.withSecondaryError (6) *secondary.withSecondaryError (7) *secondary.withSecondaryError (8) *secondary.withSecondaryError (9) *secondary.withSecondaryError (10) *secondary.withSecondaryError (11) *secondary.withSecondaryError (12) *secondary.withSecondaryError (13) *secondary.withSecondaryError (14) *secondary.withSecondaryError (15) *secondary.withSecondaryError (16) *secondary.withSecondaryError (17) *secondary.withSecondaryError (18) *secondary.withSecondaryError (19) *secondary.withSecondaryError (20) *secondary.withSecondaryError (21) *secondary.withSecondaryError (22) *secondary.withSecondaryError (23) *secondary.withSecondaryError (24) *secondary.withSecondaryError (25) *secondary.withSecondaryError (26) *secondary.withSecondaryError (27) *secondary.withSecondaryError (28) *secondary.withSecondaryError (29) *secondary.withSecondaryError (30) *secondary.withSecondaryError (31) *secondary.withSecondaryError (32) *secondary.withSecondaryError (33) *secondary.withSecondaryError (34) *secondary.withSecondaryError (35) *secondary.withSecondaryError (36) *secondary.withSecondaryError (37) *secondary.withSecondaryError (38) *secondary.withSecondaryError (39) *secondary.withSecondaryError (40) *secondary.withSecondaryError (41) *secondary.withSecondaryError (42) *secondary.withSecondaryError (43) *secondary.withSecondaryError (44) *secondary.withSecondaryError (45) *secondary.withSecondaryError (46) *secondary.withSecondaryError (47) *secondary.withSecondaryError (48) *secondary.withSecondaryError (49) *secondary.withSecondaryError (50) *secondary.withSecondaryError (51) *secondary.withSecondaryError (52) *secondary.withSecondaryError (53) *secondary.withSecondaryError (54) *secondary.withSecondaryError (55) *secondary.withSecondaryError (56) *secondary.withSecondaryError (57) *secondary.withSecondaryError (58) *secondary.withSecondaryError (59) *secondary.withSecondaryError (60) *secondary.withSecondaryError (61) *secondary.withSecondaryError (62) *secondary.withSecondaryError (63) *secondary.withSecondaryError (64) *secondary.withSecondaryError (65) *secondary.withSecondaryError (66) *secondary.withSecondaryError (67) *secondary.withSecondaryError (68) *secondary.withSecondaryError (69) *secondary.withSecondaryError (70) *secondary.withSecondaryError (71) *secondary.withSecondaryError (72) *secondary.withSecondaryError (73) *withstack.withStack (74) *errutil.withPrefix (75) *withstack.withStack (76) *errutil.leafError
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/multiregion
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*tpcc/multiregion/survive=zone/chaos=true.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-12207

roachtest.tpcc/multiregion/survive=zone/chaos=true [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4467588&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4467588&tab=artifacts#/tpcc/multiregion/survive=zone/chaos=true) on master @ [eee8ef080a95bc59eb3ddc2bebf623eb78e78db9](https://github.com/cockroachdb/cockroach/commits/eee8ef080a95bc59eb3ddc2bebf623eb78e78db9):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/tpcc/multiregion/survive=zone/chaos=true/run_1
	tpcc.go:267,tpcc.go:577,test_runner.go:779: error at from 2022-02-27T15:50:59Z, to 2022-02-27T15:55:39Z on metric workload_tpcc_newOrder_success_total{instance="35.190.174.27:2114"}: expected successes to not increase, found from 33890.000000, to 33893.000000
		(1) secondary error attachment
		  | error at from 2022-02-27T15:50:59Z, to 2022-02-27T15:55:39Z on metric workload_tpcc_stockLevel_success_total{instance="35.190.174.27:2114"}: expected successes to not increase, found from 3396.000000, to 3397.000000
		  | (1) attached stack trace
		  |   -- stack trace:
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  |   | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:208
		  |   | [...repeated from below...]
		  | Wraps: (2) error at from 2022-02-27T15:50:59Z, to 2022-02-27T15:55:39Z on metric workload_tpcc_stockLevel_success_total{instance="35.190.174.27:2114"}
		  | Wraps: (3) attached stack trace
		  |   -- stack trace:
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkDowntime.func1
		  |   | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:103
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  |   | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:207
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkDowntime
		  |   | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:92
		  |   | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).listen.func1
		  |   | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:273
		  |   | runtime.goexit
		  |   | 	GOROOT/src/runtime/asm_amd64.s:1581
		  | Wraps: (4) expected successes to not increase, found from 3396.000000, to 3397.000000
		  | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:208
		  | [...repeated from below...]
		Wraps: (3) error at from 2022-02-27T15:50:59Z, to 2022-02-27T15:55:39Z on metric workload_tpcc_newOrder_success_total{instance="35.190.174.27:2114"}
		Wraps: (4) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkDowntime.func1
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:103
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkMetrics
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:207
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).checkDowntime
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:92
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.(*tpccChaosEventProcessor).listen.func1
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/drt.go:273
		  | runtime.goexit
		  | 	GOROOT/src/runtime/asm_amd64.s:1581
		Wraps: (5) expected successes to not increase, found from 33890.000000, to 33893.000000
		Error types: (1) *secondary.withSecondaryError (2) *withstack.withStack (3) *errutil.withPrefix (4) *withstack.withStack (5) *errutil.leafError
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*tpcc/multiregion/survive=zone/chaos=true.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/cmd/roachtest/tests/drt.go
+++ b/pkg/cmd/roachtest/tests/drt.go
@@ -15,9 +15,12 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
 	"github.com/cockroachdb/cockroach/pkg/roachprod/prometheus"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/errors"
+	promv1 "github.com/prometheus/client_golang/api/prometheus/v1"
 	"github.com/prometheus/common/model"
 )
 
@@ -153,27 +156,13 @@ func (ep *tpccChaosEventProcessor) checkMetrics(
 			ep.workloadNodeIP,
 			w.prometheusPort,
 		)
-		fromVal, warnings, err := ep.promClient.Query(
-			ctx,
-			q,
-			fromTime,
-		)
+		fromVal, err := ep.queryPrometheus(ctx, l, q, fromTime)
 		if err != nil {
 			return err
 		}
-		if len(warnings) > 0 {
-			return errors.Newf("found warnings querying prometheus: %s", warnings)
-		}
-		toVal, warnings, err := ep.promClient.Query(
-			ctx,
-			q,
-			toTime,
-		)
+		toVal, err := ep.queryPrometheus(ctx, l, q, toTime)
 		if err != nil {
 			return err
-		}
-		if len(warnings) > 0 {
-			return errors.Newf("found warnings querying prometheus: %s", warnings)
 		}
 
 		// Results are a vector with 1 element, so deserialize accordingly.
@@ -209,6 +198,29 @@ func (ep *tpccChaosEventProcessor) checkMetrics(
 		}
 	}
 	return nil
+}
+
+func (ep *tpccChaosEventProcessor) queryPrometheus(
+	ctx context.Context, l *logger.Logger, q string, ts time.Time,
+) (val model.Value, err error) {
+	rOpts := base.DefaultRetryOptions()
+	rOpts.MaxRetries = 5
+	var warnings promv1.Warnings
+	for r := retry.Start(rOpts); r.Next(); {
+		val, warnings, err = ep.promClient.Query(
+			ctx,
+			q,
+			ts,
+		)
+		if err == nil {
+			break
+		}
+		l.Printf("error querying prometheus, retrying: %+v", err)
+	}
+	if len(warnings) > 0 {
+		return nil, errors.Newf("found warnings querying prometheus: %s", warnings)
+	}
+	return val, err
 }
 
 func (ep *tpccChaosEventProcessor) writeErr(ctx context.Context, l *logger.Logger, err error) {

--- a/pkg/cmd/roachtest/tests/tpcc.go
+++ b/pkg/cmd/roachtest/tests/tpcc.go
@@ -299,7 +299,7 @@ func runTPCC(ctx context.Context, t test.Test, c cluster.Cluster, opts tpccOptio
 	// Check no errors from metrics.
 	if ep != nil {
 		if err := ep.err(); err != nil {
-			t.Fatal(err)
+			t.Fatal(errors.Wrap(err, "error detected during DRT"))
 		}
 	}
 }

